File: console/src/services/api/contact_timeline.ts
import { api } from './client'

// Entity data types for each entity type
export interface ContactEntityData {
  email: string
  first_name?: string
  last_name?: string
  external_id?: string
}

export interface ContactListEntityData {
  id: string
  name: string
  status: string
}

export interface MessageHistoryEntityData {
  id: string
  template_id: string
  template_version: number
  template_name?: string | null
  template_category?: string | null
  template_email?: Record<string, any> | null
  channel: string
  sent_at: string
  delivered_at?: string
  opened_at?: string
  clicked_at?: string
  message_data?: Record<string, any>
}

export interface WebhookEventEntityData {
  id: string
  type: string // delivered, bounce, complaint, opened, clicked, auth_email, before_user_created
  source: string // ses, postmark, mailgun, sparkpost, mailjet, smtp, supabase
  message_id?: string | null
  timestamp: string
  bounce_type?: string | null
  bounce_category?: string | null
  bounce_diagnostic?: string | null
  complaint_feedback_type?: string | null
  template_id?: string | null
  template_version?: number | null
  template_name?: string | null
}

export interface ContactSegmentEntityData {
  id: string
  name: string
  color?: string
  version?: number
}

export type EntityData =
  | ContactEntityData
  | ContactListEntityData
  | MessageHistoryEntityData
  | WebhookEventEntityData
  | ContactSegmentEntityData

export interface ContactTimelineEntry {
  id: string
  email: string
  operation: 'insert' | 'update' | 'delete'
  entity_type: 'contact' | 'contact_list' | 'message_history' | 'webhook_event' | 'contact_segment'
  kind: string // operation_entityType (e.g., 'insert_contact', 'update_message_history', 'join_segment', 'leave_segment')
  changes: Record<string, any>
  entity_id?: string // NULL for contact, list_id for contact_list, message_id for message_history and webhook_event, segment_id for contact_segment
  entity_data?: EntityData // Joined entity data with contact, list, message, or webhook event details
  created_at: string // Can be set to historical data
  db_created_at: string // Timestamp when record was inserted into database
}

export interface TimelineListRequest {
  workspace_id: string
  email: string
  limit?: number // Default 50, max 100
  cursor?: string
}

export interface TimelineListResponse {
  timeline: ContactTimelineEntry[]
  next_cursor?: string
}

export const contactTimelineApi = {
  /**
   * List timeline entries for a contact with pagination
   */
  list: async (params: TimelineListRequest): Promise<TimelineListResponse> => {
    const searchParams = new URLSearchParams()

    // Add required params
    searchParams.append('workspace_id', params.workspace_id)
    searchParams.append('email', params.email)

    // Add optional params if they exist
    if (params.limit !== undefined) {
      searchParams.append('limit', params.limit.toString())
    }
    if (params.cursor) {
      searchParams.append('cursor', params.cursor)
    }

    return api.get<TimelineListResponse>(`/api/timeline.list?${searchParams.toString()}`)
  }
}

File: internal/domain/contact_timeline.go
package domain

import (
	"context"
	"fmt"
	"net/url"
	"strconv"
	"time"
)

//go:generate mockgen -destination mocks/mock_contact_timeline_service.go -package mocks github.com/Notifuse/notifuse/internal/domain ContactTimelineService
//go:generate mockgen -destination mocks/mock_contact_timeline_repository.go -package mocks github.com/Notifuse/notifuse/internal/domain ContactTimelineRepository

// ContactTimelineEntry represents a timeline entry for a contact
type ContactTimelineEntry struct {
	ID          string                 `json:"id"`
	Email       string                 `json:"email"`
	Operation   string                 `json:"operation"`   // 'insert', 'update', 'delete'
	EntityType  string                 `json:"entity_type"` // 'contact', 'contact_list', 'message_history'
	Kind        string                 `json:"kind"`        // operation_entityType (e.g., 'insert_contact', 'update_message_history')
	Changes     map[string]interface{} `json:"changes"`
	EntityID    *string                `json:"entity_id,omitempty"`   // NULL for contact, list_id for contact_list, message_id for message_history
	EntityData  map[string]interface{} `json:"entity_data,omitempty"` // Joined entity data (contact, list, or message details)
	CreatedAt   time.Time              `json:"created_at"`            // Can be set to historical data
	DBCreatedAt time.Time              `json:"db_created_at"`         // Timestamp when record was inserted into database
}

// TimelineListRequest represents the request parameters for listing timeline entries
type TimelineListRequest struct {
	WorkspaceID string
	Email       string
	Limit       int
	Cursor      *string
}

// TimelineListResponse represents the response for listing timeline entries
type TimelineListResponse struct {
	Timeline   []*ContactTimelineEntry `json:"timeline"`
	NextCursor *string                 `json:"next_cursor,omitempty"`
}

// Validate validates the timeline list request
func (r *TimelineListRequest) Validate() error {
	if r.WorkspaceID == "" {
		return fmt.Errorf("workspace_id is required")
	}
	if r.Email == "" {
		return fmt.Errorf("email is required")
	}
	if r.Limit < 0 {
		return fmt.Errorf("limit must be non-negative")
	}
	if r.Limit > 100 {
		return fmt.Errorf("limit cannot exceed 100")
	}
	return nil
}

// FromQuery parses query parameters into a TimelineListRequest
func (r *TimelineListRequest) FromQuery(query url.Values) error {
	r.WorkspaceID = query.Get("workspace_id")
	r.Email = query.Get("email")

	// Parse limit with default value
	r.Limit = 50 // Default
	if limitStr := query.Get("limit"); limitStr != "" {
		parsedLimit, err := strconv.Atoi(limitStr)
		if err != nil {
			return fmt.Errorf("invalid limit parameter: must be an integer")
		}
		r.Limit = parsedLimit
	}

	// Parse cursor
	if cursorStr := query.Get("cursor"); cursorStr != "" {
		r.Cursor = &cursorStr
	}

	return r.Validate()
}

// ContactTimelineRepository defines methods for contact timeline persistence
type ContactTimelineRepository interface {
	// Create inserts a new timeline entry
	Create(ctx context.Context, workspaceID string, entry *ContactTimelineEntry) error
	// List retrieves timeline entries for a contact
	List(ctx context.Context, workspaceID string, email string, limit int, cursor *string) ([]*ContactTimelineEntry, *string, error)
	// DeleteForEmail deletes all timeline entries for a contact
	DeleteForEmail(ctx context.Context, workspaceID string, email string) error
}

// ContactTimelineService defines business logic for contact timeline
type ContactTimelineService interface {
	// List retrieves timeline entries for a contact with pagination
	List(ctx context.Context, workspaceID string, email string, limit int, cursor *string) ([]*ContactTimelineEntry, *string, error)
}
